### PR TITLE
NEXT-12629 - Add custom SEO URL guide

### PR DESCRIPTION
**Pull request by PaddyS**

### CODE REVIEW

File: guides/plugins/plugins/content/seo/add-custom-seo-url.md [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+As already mentioned in the overview, this guide will divided into two parts: Static and dynamic SEO URLs.
+
+### Static SEO URLs
+
+A static SEO URL doesn't have to change every now and then.
+Imagine a custom controller, which is accessible via the link `yourShop.com/example`.
+
+Now if you want this URL to be translatable, you'll have to add a custom SEO URL to your controller route,
+so it is accessible using both `Example-Page` in english, as well as e.g. `Beispiel-Seite` in german.
```
Comment (JoshuaBehrens) [inline guides/plugins/plugins/content/seo/add-custom-seo-url.md:35]: english -> English
germain -> German

File: guides/plugins/plugins/content/seo/add-custom-seo-url.md [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+{% endcode %}
+
+The important information you'll need here are the route name, `frontend.example.example`, as well as the route itself: `/example`.
+Make sure to remember those for the next step.
+
+#### Example migration
+
+Creating a SEO URL in this scenario can be achieved by creating a [plugin migration](../../plugin-fundamentals/database-migrations.md).
+
+The migration has to insert an entry for each sales-channel and language into the `seo_url` table.
```
Comment (JoshuaBehrens) [inline guides/plugins/plugins/content/seo/add-custom-seo-url.md:81]: sales-channel -> sales channel, right @leichteckig ?

File: guides/plugins/plugins/content/seo/add-custom-seo-url.md [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+For thise case, we're making use of the `ImportTranslationsTrait`, which comes with a helper method `importTranslation`.
+
+Don't be confused here, we'll just treat the `seo_url` table like a translation table, since it also needs a `language_id`
+and respective translated SEO URLs.
+You'll have to pass a german and an english array into an instance of the `Translations` class, which is then the second
```
Comment (JoshuaBehrens) [inline guides/plugins/plugins/content/seo/add-custom-seo-url.md:86]: german -> German
english -> English

File: guides/plugins/plugins/content/seo/add-custom-seo-url.md [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+You might want to have a look at the `getSeoMetaArray` method, that we implemented here.
+Most important for you are the columns `route_name` and `path_info` here, which represent the values you've defined in
+your controller's route annotation.
+
+By using the default PHP method `array_merge`, we're then also adding our translated SEO URL to the column `seo_path_info`.
+
+And that's it! After installing our plugin, you should now be able to access your controller's route with the given SEO URLs.
+
+{% hint style="info"%}
+You can only access the german SEO URL if you've configured a german domain in your respective sales channel first.
```
Comment (JoshuaBehrens) [inline guides/plugins/plugins/content/seo/add-custom-seo-url.md:172]: german -> German

**Review by JoshuaBehrens — CHANGES_REQUESTED** [review]

File: guides/plugins/plugins/content/seo/add-custom-seo-url.md (hunk 1 above)
Comment (PaddyS) [reply guides/plugins/plugins/content/seo/add-custom-seo-url.md:35]: You sure about that? I know it's the case for the countries, as in "Germany" - but also for the language?

File: guides/plugins/plugins/content/seo/add-custom-seo-url.md [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+
+And that's it! After installing our plugin, you should now be able to access your controller's route with the given SEO URLs.
+
+{% hint style="info"%}
+You can only access the German SEO URL if you've configured a German domain in your respective sales channel first.
+{% endhint %}
+
+### Dynamic SEO URLs
+
+Dynamic SEO URLs are URLs, that have to change every now and then.
```
Comment (PaddyS) [inline guides/plugins/plugins/content/seo/add-custom-seo-url.md:177]: Change this to non-entity examples. Use e.g. `ProductPageSeoUrlRoute` instead for entity examples

Comment (PaddyS) [reply guides/plugins/plugins/content/seo/add-custom-seo-url.md:177]: Also add: Must execute `SeoUrlUpdater::update` on written event still when using the PageSeourlRoute class

File: guides/plugins/plugins/content/seo/add-custom-seo-url.md [hunk 6]
```diff
@@ -0,0 +1,356 @@
+# Add custom SEO URL
+
+## Overview
+
+Every good website had to deal with it already: SEO URLs.
```
Comment (lernhart) [inline guides/plugins/plugins/content/seo/add-custom-seo-url.md:5]: Did you mean sth like this ?

```suggestion
Every good website had to deal with it at some point: SEO URLs.
```

File: guides/plugins/plugins/content/seo/add-custom-seo-url.md [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+Furthermore, we're going to use a [custom storefront controller](../../storefront/add-custom-controller.md) for the static SEO URL example,
+as well as [custom entities](../../framework/data-handling/add-custom-complex-data.md) for the dynamic SEO URLs.
+Make sure you know and understand those two as well before diving deeper into this guide.
+Those come with two different solutions:
+- Using [plugin migrations](../../plugin-fundamentals/database-migrations.md) for static SEO URLs
+- Using [DAL events](../../framework/data-handling/using-database-events.md) to react on entity changes and therefore generating a dynamic SEO URL
+
+## Custom SEO URLs
+
+As already mentioned in the overview, this guide will divided into two parts: Static and dynamic SEO URLs.
```
Comment (lernhart) [inline guides/plugins/plugins/content/seo/add-custom-seo-url.md:27]:
```suggestion
As already mentioned in the overview, this guide is divided into two parts: Static and dynamic SEO URLs.
```

File: guides/plugins/plugins/content/seo/add-custom-seo-url.md [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+Now if you want this URL to be translatable, you'll have to add a custom SEO URL to your controller route,
+so it is accessible using both `Example-Page` in English, as well as e.g. `Beispiel-Seite` in German.
+
+#### Example controller
+
+For this example, the controller from the [Add custom controller guide](../../storefront/add-custom-controller.md) is being used.
+It creates a controller with a route like the example mentioned above: `/example`
+
+Let's now have a look at our example controller:
+{% code title="<plugin root>/src/Migration/Migration1619094740AddStaticSeoUrl.php" %}
```
Comment (lernhart) [inline guides/plugins/plugins/content/seo/add-custom-seo-url.md:43]:
```suggestion
{% code title="<plugin root>/src/Storefront/Controller/ExampleController.php" %}
```

File: guides/plugins/plugins/content/seo/add-custom-seo-url.md [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+    {
+        return $this->renderStorefront('@SwagBasicExample/storefront/page/example/index.html.twig', [
+            'example' => 'Hello world'
+        ]);
+    }
+}
+```
+{% endcode %}
+
+The important information you'll need here are the route name, `frontend.example.example`, as well as the route itself: `/example`.
```
Comment (lernhart) [inline guides/plugins/plugins/content/seo/add-custom-seo-url.md:74]:
```suggestion
The important information you'll need here is the route name, `frontend.example.example`, as well as the route itself: `/example`.
```

**Review by lernhart — CHANGES_REQUESTED** [review]

**Pull request merged**